File: ocr_service.py
"""
OCR service for extracting card names from images.
Uses Tesseract OCR to identify card text.
"""
import cv2
import numpy as np
from typing import Optional, List
try:
    import pytesseract
    TESSERACT_AVAILABLE = True
except ImportError:
    TESSERACT_AVAILABLE = False
    print("Warning: pytesseract not available. OCR functionality will be limited.")


class OCRService:
    """Service for extracting text from card images using OCR."""
    
    def __init__(self):
        """Initialize OCR service."""
        self.tesseract_available = TESSERACT_AVAILABLE
    
    def preprocess_image(self, image: np.ndarray) -> np.ndarray:
        """
        Preprocess image for better OCR results.
        
        Args:
            image: OpenCV image
            
        Returns:
            Preprocessed image
        """
        # Convert to grayscale
        if len(image.shape) == 3:
            gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
        else:
            gray = image.copy()
        
        # Apply thresholding to get better text contrast
        _, thresh = cv2.threshold(gray, 0, 255, cv2.THRESH_BINARY + cv2.THRESH_OTSU)
        
        # Denoise
        denoised = cv2.fastNlMeansDenoising(thresh)
        
        return denoised
    
    def extract_card_name_from_region(self, image: np.ndarray, top_percentage: float = 0.15) -> Optional[str]:
        """
        Extract card name from the top region of a card image.
        MTG card names are typically in the top 15% of the card.
        
        Args:
            image: OpenCV image of the card
            top_percentage: Percentage of card height to consider for name
            
        Returns:
            Extracted card name or None
        """
        if not self.tesseract_available:
            return None
        
        try:
            height, width = image.shape[:2]
            name_region_height = int(height * top_percentage)
            
            # Extract the top region where the card name should be
            name_region = image[0:name_region_height, :]
            
            # Preprocess the region
            processed = self.preprocess_image(name_region)
            
            # Perform OCR
            text = pytesseract.image_to_string(processed, config='--psm 7')
            
            # Clean up the text
            card_name = text.strip()
            
            # Remove common OCR artifacts and non-alphanumeric characters except spaces and commas
            card_name = ''.join(c for c in card_name if c.isalnum() or c in [' ', ',', "'", '-'])
            card_name = ' '.join(card_name.split())  # Normalize whitespace
            
            if card_name and len(card_name) > 2:
                return card_name
            
            return None
            
        except Exception as e:
            print(f"Error extracting card name: {e}")
            return None
    
    def extract_text_from_image(self, image_path: str) -> List[str]:
        """
        Extract all text from an image.
        
        Args:
            image_path: Path to image file
            
        Returns:
            List of extracted text lines
        """
        if not self.tesseract_available:
            return []
        
        try:
            # Read image
            image = cv2.imread(image_path)
            if image is None:
                return []
            
            # Preprocess
            processed = self.preprocess_image(image)
            
            # Perform OCR
            text = pytesseract.image_to_string(processed)
            
            # Split into lines and clean
            lines = [line.strip() for line in text.split('\n') if line.strip()]
            
            return lines
            
        except Exception as e:
            print(f"Error extracting text: {e}")
            return []
    
    def is_available(self) -> bool:
        """Check if OCR functionality is available."""
        return self.tesseract_available

File: ocr_service.rb
# frozen_string_literal: true

# OCRService - Service for extracting text from card images using OCR
class OCRService
  def initialize
    @tesseract_available = check_tesseract
  end

  # Check if Tesseract OCR is available
  # @return [Boolean] True if Tesseract is installed
  def check_tesseract
    system('which tesseract > /dev/null 2>&1')
  end

  # Extract card name from the top region of a card image
  # MTG card names are typically in the top 15% of the card
  # @param image_path [String] Path to the card image
  # @param top_percentage [Float] Percentage of card height to consider for name
  # @return [String, nil] Extracted card name or nil
  def extract_card_name_from_region(image_path, top_percentage = 0.15)
    return nil unless @tesseract_available

    begin
      # Use Tesseract to extract text from image
      # Using --psm 7 for single line of text
      text = `tesseract "#{image_path}" stdout --psm 7 2>/dev/null`.strip
      
      # Clean up the text
      card_name = text.gsub(/[^a-zA-Z0-9\s,'\-]/, '').squeeze(' ').strip
      
      return card_name if card_name.length > 2
      
      nil
    rescue StandardError => e
      puts "Error extracting card name: #{e.message}"
      nil
    end
  end

  # Extract all text from an image
  # @param image_path [String] Path to image file
  # @return [Array<String>] List of extracted text lines
  def extract_text_from_image(image_path)
    return [] unless @tesseract_available

    begin
      text = `tesseract "#{image_path}" stdout 2>/dev/null`
      text.split("\n").map(&:strip).reject(&:empty?)
    rescue StandardError => e
      puts "Error extracting text: #{e.message}"
      []
    end
  end

  # Check if OCR functionality is available
  # @return [Boolean] True if Tesseract is available
  def available?
    @tesseract_available
  end
end
